src/riak_ensemble_peer.erl

```diff
@@ -1884,22 +1902,29 @@ handle_sync_event(tree_info, _From, StateName, State=#state{tree_trust=Trust,
 handle_sync_event({debug_local_get, Key}, From, StateName, State) ->
     State2 = do_local_get(From, Key, State),
     {next_state, StateName, State2};
+handle_sync_event(get_watchers, _From, StateName, State) ->
+    {reply, State#state.watchers, StateName, State};
 handle_sync_event(_Event, _From, StateName, State) ->
     Reply = ok,
     {reply, Reply, StateName, State}.
 
 %% -spec handle_info(_, atom(), state()) -> next_state().
-handle_info({'DOWN', Ref, _, Pid, Reason}, StateName,
-  #state{mod=Mod, modstate=ModState}=State) ->
-    case Mod:handle_down(Ref, Pid, Reason, ModState) of
+handle_info({'DOWN', MRef, _, Pid, Reason}, StateName, State) ->
+    #state{mod=Mod, modstate=ModState, watchers=Watchers} = State,
+    case lists:keytake(Pid, 1, Watchers) of
```

Comment:
not crazy about the nested case statements here - can we pull the `lists:keytake` part into a `remove_watcher` function that takes the pid and either returns a new list or `not_found` if it's not there? Also, in the `not_found` case I'd just call another function that encompasses the Mod:handle_down stuff - maybe `proxy_handle_down` or `mod_handle_down` or something to that effect? would make undertanding what's going on here easier I think.

I'd also add a quick comment explaining why there are two different ways we are dealing with `DOWN` messages.